python/llm/example/GPU/LLM-Finetuning/LoRA/lora_finetuning.py

```diff
@@ -0,0 +1,405 @@
+            f"lora_target_modules: {lora_target_modules}\n"
+            f"train_on_inputs: {train_on_inputs}\n"
+            f"add_eos_token: {add_eos_token}\n"
+            f"group_by_length: {group_by_length}\n"
+            f"wandb_project: {wandb_project}\n"
+            f"wandb_run_name: {wandb_run_name}\n"
+            f"wandb_watch: {wandb_watch}\n"
+            f"wandb_log_model: {wandb_log_model}\n"
+            f"resume_from_checkpoint: {resume_from_checkpoint or False}\n"
+            f"prompt template: {prompt_template_name}\n"
+            f"training_mode: {training_mode}\n"
+            f"relora_steps: {relora_steps}\n"
+            f"relora_warmup_steps: {relora_warmup_steps}\n"
+            f"relora_cpu_offload: {relora_cpu_offload}\n"
+        )
+    assert (
+        base_model
+    ), "Please specify a --base_model, e.g. --base_model='huggyllama/llama-7b'"
+    gradient_accumulation_steps = batch_size // micro_batch_size
+
+    if training_mode == "relora":
+        assert(relora_steps > 0), "The relora_steps should > 0 if the training_mode is relora."
+
+    prompter = Prompter(prompt_template_name)
+
+    device_map = "auto"
+    world_size = int(os.environ.get("WORLD_SIZE", 1))
+    ddp = world_size != 1
+    if ddp:
+        device_map = {"": int(os.environ.get("LOCAL_RANK") or 0)}
+        gradient_accumulation_steps = gradient_accumulation_steps // world_size
+
+    # Check if parameter passed or if set within environ
+    use_wandb = len(wandb_project) > 0 or (
+        "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
+    )
+    # Only overwrite environ if wandb param passed
+    if len(wandb_project) > 0:
+        os.environ["WANDB_PROJECT"] = wandb_project
+    if len(wandb_watch) > 0:
+        os.environ["WANDB_WATCH"] = wandb_watch
+    if len(wandb_log_model) > 0:
+        os.environ["WANDB_LOG_MODEL"] = wandb_log_model
+
+    if saved_low_bit_model is not None:
+        # Load the low bit optimized model if provide the saved path
+        model = AutoModelForCausalLM.load_low_bit(
+            saved_low_bit_model,
+            optimize_model=False,
+            torch_dtype=torch.bfloat16,
+            modules_to_not_convert=["lm_head"],
+        )
+    else:
+        # According to the QLoRA paper, using "nf4" could yield better model quality than "int4"
+        # Default 4-bit format for qa-lora is sym_int4
+        if training_mode == "lora":
+            model = AutoModelForCausalLM.from_pretrained(
+                base_model,
+                load_in_low_bit="bf16",
+                optimize_model=False,
+                torch_dtype=torch.bfloat16,
+                modules_to_not_convert=["lm_head"],
+            )
+        else:
+            # use bnb_config for qlora/qalora/relora, which use 4bit for base model
+            if training_mode == "qalora":
+                low_bit_format = "int4"
+            else:
+                low_bit_format = "nf4"
```

Comment:
We should simply this and only support LoRA here, and refactor the common code across QLoRA, etc., to `common`

Reply:
> We should simply this and only support LoRA here, and refactor the common code across QLoRA, etc., to `common`

Got it, I'll extract common code to `common`.